src/runtime_src/core/include/experimental/xrt_xclbin.h

```diff
@@ -33,13 +36,270 @@
+     */
+    XCL_DRIVER_DLLESPEC
+    int32_t
+    get_index() const;
+  };
+
+  /**
+   * class arg - xrt::xclbin::arg represents a kernel or cu argument
+   *
+   * Kernels must be provided arguments to operate on
+   *   This class provides meta-data associated with a given argument
+   */
+  class arg_impl;
+  class arg : public Pimpl<arg_impl>
+  {
+  public:
+    explicit arg(std::shared_ptr<arg_impl> handle) : Pimpl(std::move(handle)) {}
+
+    /**
+     * get_name() - Get argument name
+     *
+     * Return: argument name
+     *
+     */
+    XCL_DRIVER_DLLESPEC
+    const std::string&
+    get_name() const;
+
+    /**
+     * get_mems() - Get list of device memories from xclbin.
+     *
+     * Return: A list of memories from xclbin.
+     */
+    XCL_DRIVER_DLLESPEC
+    std::vector<mem>
+    get_mems() const;
+
+    /**
+     * get_mems() - Get a memory by index from an xclbin.
+     *
+     * Return: A mem from xclbin.
+     */
+    XCL_DRIVER_DLLESPEC
+    mem
+    get_mems(int32_t index) const;
+  };
+
+  /**
+   * class cu - xrt::xclbin::cu represents a compute unit in an xclbin
+   *
+   * The cu handle is explicitly constructed by the xclbin object.
+   *
+   */
+  class cu_impl;
+  class cu : public Pimpl<cu_impl>
+  {
+  public:
+    explicit cu(std::shared_ptr<cu_impl> handle) : Pimpl(std::move(handle)) {}
+
+    /**
+     * get_name() - Get CU name
+     *
+     * Return: CU Name
+     *
+     */
+    XCL_DRIVER_DLLESPEC
+    const std::string&
+    get_name() const;
+
+    /**
+     * get_args() - Get list of arguments associated with this object.
+     *
+     * Return: A list of arguments.
+     */
+    XCL_DRIVER_DLLESPEC
+    std::vector<arg>
+    get_args() const;
+
+    /**
+     * get_args() - Get argument at index.
+     *
+     * Return: A list of arguments.
+     */
+    XCL_DRIVER_DLLESPEC
+    arg
+    get_args(int32_t index) const;
+
+    /**
+     * get_base_address() - Get the base address of the cu
+     *
+     * Return: Base address of the cu
+     *
+     */
+    XCL_DRIVER_DLLESPEC
+    uint64_t
+    get_base_address() const;
+  };
+
+  /**
+   * class kernel - xrt::kernel represents a kernel in an xclbin
+   *
+   * The kernel handle is explicitly constructed by the xclbin object.
+   *
+   */
+  class kernel_impl;
+  class kernel : public Pimpl<kernel_impl>
+  {
+  public:
+    explicit kernel(std::shared_ptr<kernel_impl> handle) : Pimpl(std::move(handle)) {}
+
+    /**
+     * get_name() - Get kernel name
+     *
+     * Return: kernel name
+     *
+     */
+    XCL_DRIVER_DLLESPEC
+    const std::string&
+    get_name() const;
+
+    /**
+     * get_cus() - Get list of cu from kernel.
+     *
+     * Return: A list of cu from kernel.
+     */
+    XCL_DRIVER_DLLESPEC
+    std::vector<cu>
+    get_cus() const;
+
+    /**
+     * get_cus() - Get cu from kernel at index.
+     *
+     * Return: A cu from kernel at index.
+     */
+    XCL_DRIVER_DLLESPEC
+    cu
+    get_cus(int32_t index) const;
```

Comment:
Should not be plural when single is returned.    What does the index actually signify?  There is no such concept of indexning a kernel for a CU?